### PR TITLE
add ut of ads_event.go

### PR DESCRIPTION
**Pull request by LiZhenCheng9527**
**What type of PR is this?**
/kind enhancement

**What this PR does / why we need it**:
add ut of ads_event.go
**Which issue(s) this PR fixes**:
a part of #104

### CODE REVIEW

File: pkg/controller/envoy/ads_event_test.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+		assert.Equal(t, core_v2.ApiStatus_ALL, apiMethod)
+		resourceName := sets.New[string]()
+		resourceName.Insert(listener.GetName())
+		diff := svc.DynamicLoader.ListenerCache.GetResourceNames().Difference(resourceName)
+		result := sets.Set[string]{}
+		assert.Equal(t, result, diff)
+	})
+}
+
+func TestHandleRdaResponse(t *testing.T) {
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:297]: typo

File: pkg/controller/envoy/ads_event_test.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+					anyCluster,
+				},
+			}
+			err = tt.svc.handleCdsResponse(rsp)
+			assert.NoError(t, err)
+			assert.Equal(t, tt.wantClusterName, tt.svc.DynamicLoader.edsClusterNames)
+			resourceName := sets.New[string]()
+			resourceName.Insert(tt.cluster.GetName())
+			diff := tt.svc.DynamicLoader.ClusterCache.GetResourceNames().Difference(resourceName)
+			assert.Equal(t, diff, tt.wantSet)
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:101]: what does this test?

**Review by hzxuzhonghu — COMMENTED** [review]
can you test more cases with different clusters

File: pkg/controller/envoy/ads_event_test.go [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("test3: nil cluster", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{}
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:95]: what does nil cluster mean? how could it happen

File: pkg/controller/envoy/ads_event_test.go [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_LOGICAL_DNS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		clusterHash := hash.Sum64String(anyCluster.String())
+		svc.DynamicLoader.ClusterCache.SetCdsHash(cluster.GetName(), clusterHash)
+		apiStatus := core_v2.ApiStatus_UPDATE
+		svc.DynamicLoader.CreateApiClusterByCds(apiStatus, cluster)
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:127]: replace with calling handleCdsResponse

File: pkg/controller/envoy/ads_event_test.go [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+
+		cluster.Name = "cluster"
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:129]: I think this is not updating cluster, this is remove a old one, and then add a new one.

As multi clusters can be returned, the test case multiplexed

File: pkg/controller/envoy/ads_event_test.go [hunk 6]
```diff
@@ -0,0 +1,637 @@
...
+			anyCluster, err := anypb.New(cluster)
+			assert.NoError(t, err)
+			clusterHash := hash.Sum64String(anyCluster.String())
+			svc.DynamicLoader.ClusterCache.SetCdsHash(cluster.GetName(), clusterHash)
+			apiStatus := core_v2.ApiStatus_UPDATE
+			svc.DynamicLoader.CreateApiClusterByCds(apiStatus, cluster)
+			svc.DynamicLoader.edsClusterNames = []string{
+				"ut-cluster2",
+			}
+		}
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:174]: replace with calling handler

File: pkg/controller/envoy/ads_event_test.go [hunk 7]
```diff
@@ -0,0 +1,637 @@
...
+		err = svc.handleLdsResponse(rsp)
+		assert.NoError(t, err)
+		apiMethod := svc.DynamicLoader.ListenerCache.GetApiListener("ut-listener").ApiStatus
+		assert.Equal(t, core_v2.ApiStatus_UPDATE, apiMethod)
+		wantHash := hash.Sum64String(anyListener.String())
+		actualHash := svc.DynamicLoader.ListenerCache.GetLdsHash(listener.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("test2: listenerCache already has resource", func(t *testing.T) {
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:445]: can remove`test2` prefix as it is already unordered now

**Comment by LiZhenCheng9527** [comment]
This PR may need to be modified based on the results of the issues #190

File: pkg/controller/envoy/ads_event_test.go [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("already have cluster, change cluster hash", func(t *testing.T) {
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:90]: cluster update case

File: pkg/controller/envoy/ads_event_test.go [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+
+		cluster.AltStatName = "ut-name"
+		anyCluster, err = anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp = &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
```
Comment (hzxuzhonghu) [inline pkg/controller/envoy/ads_event_test.go:119]: why donot you need to mock the map r/w  like maps_v2.ClusterUpdatep

Comment (LiZhenCheng9527) [reply pkg/controller/envoy/ads_event_test.go:119]: maps_v2.ClusterUpdate handles maps in the kerne. And I dont use ApiStatu as a basis for testing. So I dont mock maps_v2.ClusterUpdatep.

Comment (LiZhenCheng9527) [reply pkg/controller/envoy/ads_event_test.go:119]: I will add testcase of Flush() in ut of cache/v2/cluster_test.go to cover usecases of handleCdsResponse

Comment (hzxuzhonghu) [reply pkg/controller/envoy/ads_event_test.go:119]: UT is to simulate real sceanrio, i am not sure how you make it no bpf map write when a new cluster  added

File: pkg/controller/envoy/ads_event_test.go [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+		}
+		err = svc.handleRdsResponse(rsp)
+		assert.NoError(t, err)
+		wantHash := hash.Sum64String(anyRouteConfig.String())
+		actualHash := svc.DynamicLoader.RouteCache.GetRdsHash(routeConfig.GetName())
+		assert.Equal(t, wantHash, actualHash)
+		assert.Equal(t, []string{"ut-routeclient", "ut-routeconfig"}, svc.ack.ResourceNames)
+	})
+
+	t.Run("nil routeConfig", func(t *testing.T) {
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event_test.go:662]: name it "empty routeConfig" would be better, `routeConfig` is not nil anyway.

File: pkg/controller/envoy/ads_event_test.go [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+		}
+		err = svc.handleRdsResponse(rsp)
+		assert.NoError(t, err)
+		wantHash := hash.Sum64String(anyRouteConfig.String())
+		actualHash := svc.DynamicLoader.RouteCache.GetRdsHash(routeConfig.GetName())
+		assert.Equal(t, wantHash, actualHash)
+		assert.Equal(t, []string{"ut-routeclient"}, svc.ack.ResourceNames)
+	})
+
+	t.Run("already have a Rda, RdaHash has change", func(t *testing.T) {
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event_test.go:688]: What does "Rda" means?

File: pkg/controller/envoy/ads_event_test.go [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+		svc.ack = newAckRequest(rsp)
+		err = svc.handleRdsResponse(rsp)
+		assert.NoError(t, err)
+		wantHash := hash.Sum64String(anyRouteConfig.String())
+		actualHash := svc.DynamicLoader.RouteCache.GetRdsHash(routeConfig.GetName())
+		assert.Equal(t, wantHash, actualHash)
+		assert.Equal(t, []string{"ut-routeconfig"}, svc.ack.ResourceNames)
+	})
+
+	t.Run("already have a Rda, RdaHash has change. And have mult routeconfig in rsp", func(t *testing.T) {
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event_test.go:733]: `s/Rda/Rds/`,  `s/mult/multi/`, `s/rsp/resp/`, please check all the code to make sure all the typos are fixed.

File: pkg/controller/envoy/ads_event_test.go [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+		actualOldClusterHash1 := svc.DynamicLoader.ClusterCache.GetCdsHash(multClusters[0].GetName())
+		assert.Equal(t, wantOldClusterHash1, actualOldClusterHash1)
+		wantOldClusterHash2 := hash.Sum64String(anyMultCluster2.String())
+		actualOldClusterHash2 := svc.DynamicLoader.ClusterCache.GetCdsHash(multClusters[1].GetName())
+		assert.Equal(t, wantOldClusterHash2, actualOldClusterHash2)
+		assert.Equal(t, sets.Set[string]{}, clusterNames)
+		assert.Equal(t, []string{"new-ut-cluster"}, svc.rqt.ResourceNames)
+	})
+
+	t.Run("mult clusters in resp", func(t *testing.T) {
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event_test.go:231]: 's/mult/multi/'

File: pkg/controller/envoy/ads_event_test.go [hunk 14]
```diff
@@ -0,0 +1,792 @@
...
+		assert.NoError(t, err)
+		assert.Equal(t, []string{}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+		assert.Equal(t, sets.Set[string]{"ut-cluster": sets.Empty{}}, clusterNames)
+		assert.Nil(t, svc.rqt)
+	})
+
+	t.Run("have multiClusters, add a new  eds cluster", func(t *testing.T) {
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event_test.go:162]: delete the extra space between 'new' and 'eds'

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**